### PR TITLE
Add support for ExpansionPanel custom splash color

### PR DESCRIPTION
**Pull request by benji-farquhar**
Relates to https://github.com/flutter/flutter/pull/147098 and https://github.com/flutter/flutter/issues/147097

Aside from fixing the expland/collapse icon color in the other PR, I noticed the splash color on both the icon button and the full expansion panel (if ExpansionPanel.canTapOnHeader is set to true) is just `Theme.of(context).splashColor` on the `InkWell` and `Theme.of(context).highlightColor` on the `IconButton` which may not suit the color scheme of the `ExpansionPanel`, so I have added a custom field `splashColor`, which will effect both the `IconButton` and the full panel `Inkwell`. 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

File: packages/flutter/lib/src/material/expansion_panel.dart [hunk 1]
```diff
@@ -95,6 +96,11 @@ class ExpansionPanel {
   /// Defaults to false.
   final bool isExpanded;
 
+  /// Defines the splash color of the panel if [canTapOnHeader] is true, or the expand/collapse IconButton if [canTapOnHeader] is false.
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/material/expansion_panel.dart:99]:
```suggestion
  /// Defines the splash color of the panel if [canTapOnHeader] is true, or the splash color of the expand/collapse IconButton if [canTapOnHeader] is false.
```

File: packages/flutter/lib/src/material/expand_icon.dart [hunk 2]
```diff
@@ -94,6 +95,11 @@ class ExpandIcon extends StatefulWidget {
   /// and for [dark theme](https://material.io/design/color/dark-theme.html#ui-application)
   final Color? expandedColor;
 
+  /// Defines the splash color of the IconButton.
+  ///
+  /// Defaults to the Theme's highlight color, [ThemeData.highlightColor].
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/material/expand_icon.dart:100]:
```suggestion
  /// Defaults to [ThemeData.highlightColor].
```

**Review by dkwingsmt — COMMENTED** [review]
To reconfirm, if  `canTapOnHeader` is true, is there no way to trigger the icon splash at all?

Also, can you add tests?

File: packages/flutter/lib/src/material/expansion_panel.dart (hunk 1 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/material/expansion_panel.dart:99]: It's a bit hard to understand the reasoning at first glance though. Maybe:
```suggestion
	  /// Defines the splash color of the expand/collapse icon button, and the splash color of the panel if [canTapOnHeader] is true.
```

File: packages/flutter/lib/src/material/expansion_panel.dart [hunk 3]
```diff
@@ -95,6 +96,11 @@ class ExpansionPanel {
...
+  ///
+  /// Defaults to [ThemeData.splashColor] if [canTapOnHeader] is true, or [ThemeData.highlightColor] if [canTapOnHeader] is false.
```
[the remark this answers is not in the document]
Comment (BenjiFarquhar) [reply packages/flutter/lib/src/material/expansion_panel.dart:101]: @dkwingsmt  The grammar change is fine, but you have swapped the splashColor and highlightColor around to the opposite. The splashColor is in effect for the panel and the highlightColor is in effect for the IconButton.

**Comment by BenjiFarquhar** [comment]
> To reconfirm, if `canTapOnHeader` is true, is there no way to trigger the icon splash at all?
> 
> Also, can you add tests?

Correct. The Icon shouldn't splash if `canTapOnHeader` is true, as the whole panel splashes, so it would look too busy. It is more of a signal to the user that the panel is expandable if `canTapOnHeader` is true. 

I will add tests.

File: packages/flutter/lib/src/material/expansion_panel.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (dkwingsmt) [reply packages/flutter/lib/src/material/expansion_panel.dart]: Thanks!

File: packages/flutter/lib/src/material/expansion_panel.dart [hunk 4]
```diff
@@ -95,6 +96,12 @@ class ExpansionPanel {
   /// Defaults to false.
   final bool isExpanded;
 
+  /// Defines the splash color of the panel if [canTapOnHeader] is true, or the splash color of the expand/collapse IconButton if [canTapOnHeader] is false.
+  ///
+  /// If this is null, then the icon button will use its default splash color [ThemeData.highlightColor], and the panel will
+  /// use its default splash color [ThemeData.splashColor] (if [canTapOnHeader] is true).
+  final Color? splashColor;
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/material/expansion_panel.dart:103]: nit: Can you wrap the paragraphs at around 80 characters?

**Comment by BenjiFarquhar** [comment]
@dkwingsmt I have implemented the changes, I noticed `highlightColor` was available for `InkWell` and `IconButton` so I just added them both, it is a bit more clear than setting the `IconButton.highlightColor` to `ExpansionPanel.splashColor`. 

I might have exposed a bug with `IconButton.splashColor`. We don't see the blue it in the below example, but the widget test that checks the `IconButton.splashColor` passes. Not sure if it is a bug but we can see the blue `Inkwell.splashColor` very visibly, unlike `IconButton.splashColor`. But for the purposes of this specific feature, it is done unless you see any issues.

You can test with this in `main.dart`:

```
import 'package:flutter/material.dart';

/// Flutter code sample for [ExpansionPanelList].

void main() => runApp(const ExpansionPanelListExampleApp());

class ExpansionPanelListExampleApp extends StatelessWidget {
  const ExpansionPanelListExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: const Text('ExpansionPanelList Sample')),
        body: const ExpansionPanelListExample(),
      ),
    );
  }
}

// stores ExpansionPanel state information
class Item {
  Item({
    required this.expandedValue,
    required this.headerValue,
    required this.value,
    required this.canTapOnHeader,
  });

  String expandedValue;
  String headerValue;
  int value;
  bool canTapOnHeader;
}

List<Item> generateItems(int numberOfItems) {
  return List<Item>.generate(numberOfItems, (int index) {
    bool canTapOnHeader = index > 3;
    return Item(
      headerValue: 'Panel $index canTapOnHeader=${canTapOnHeader.toString()}',
      expandedValue: 'This is item number $index',
      value: index,
      canTapOnHeader: canTapOnHeader,
    );
  });
}

class ExpansionPanelListExample extends StatefulWidget {
  const ExpansionPanelListExample({super.key});

  @override
  State<ExpansionPanelListExample> createState() =>
      _ExpansionPanelListExampleState();
}

class _ExpansionPanelListExampleState extends State<ExpansionPanelListExample> {
  final List<Item> _data = generateItems(8);

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Container(
        child: _buildPanel(),
      ),
    );
  }

  Widget _buildPanel() {
    return ExpansionPanelList.radio(
      children: _data.map<ExpansionPanelRadio>((Item item) {
        return ExpansionPanelRadio(
          splashColor: Colors.blue,
          highlightColor: Colors.green,
          headerBuilder: (BuildContext context, bool isExpanded) {
            return ListTile(
              title: Text(item.headerValue),
            );
          },
          body: ListTile(
              title: Text(item.expandedValue),
          ),
          value: item.value,
          canTapOnHeader: item.canTapOnHeader,
        );
      }).toList(),
      expandIconColor: Colors.red,
    );
  }
}
```

**Comment by BenjiFarquhar** [comment]
Hi @dkwingsmt, can we progress this? It should be visibly working by running the example code in my above comment, and the widget tests verify the colors get passed through to the underlying widgets.

**Review by dkwingsmt — APPROVED** [review]
LGTM

**Comment by dkwingsmt** [comment]
The change looks great! However I don't quite understand the comment about the bug you revealed. Is it fixed with this PR? Or do you mean that even though this PR passes the value to `IconButton`, the value is not correctly applied, and this needs to be addressed in the future?

**Comment by BenjiFarquhar** [comment]
@dkwingsmt Yeah `IconButton.splashColor` is broken. I didn't fix it. I'm trying to swap out the `ExpandIcon` for the `IconButton` in the test to prove that `IconButton` has the correct `splashColor` and it just doesn't display, but something fishy is going on with Flutter master... Does `flutter pub get` work for you when you get latest?

**Comment by BenjiFarquhar** [comment]
Thanks @dkwingsmt. I have updated the test to verify that the underlying `IconButton.splashColor` and `IconButton.highlightColor` are both correctly set to the `ExpansionPanel` equivalent fields; `ExpansionPanel.splashColor` and `ExpansionPanel.highlightColor`, as this is the widget that will display the button splashes. 

There was some confusion around the fact that `IconButton.splashColor` is unused in material 3. From the `IconButton.splashColor` code comments:

> If [ThemeData.useMaterial3] is set to true, this will not be used. Use [highlightColor] instead to show the overlay color of the button when the button is in the pressed state.

However, `Inkwell` does use `splashColor` in Material 3, so both properties are useful, no matter which version of Material theme is used. 

I have updated the doc comments to inform the developer when `highlightColor` should be used instead of `splashColor`.

**Review by dkwingsmt — APPROVED** [review]
LGTM! Great finding!

**Review by justinmc — APPROVED** [review]
LGTM :+1:

**Comment by BenjiFarquhar** [comment]
Cheers @dkwingsmt & @justinmc. Good to merge?

**Pull request merged**